src/main/java/fi/aalto/cs/apluscourses/ui/exercise/ExercisesView.java

```diff
@@ -44,6 +48,14 @@ private void createUIComponents() {
     exerciseGroupsTree.setCellRenderer(new ExercisesTreeRenderer());
     exerciseGroupsTree.addNodeAppliedListener(
         ActionUtil.createOnEventLauncher(OpenItemAction.ACTION_ID, exerciseGroupsTree));
+
+    new TreeSpeedSearch(exerciseGroupsTree, x -> {
+      Object treeObject = ((DefaultMutableTreeNode) x.getLastPathComponent()).getUserObject();
```

Comment:
It's a bit unfortunate that the code here has to deal with `DefaultMutableTreeNode`, since `TreeView` doesn't really "officially promise" to use `DefaultMutableTreeNode`.

One option would be that `TreeSpeedSearch` is instead added in `TreeView`, and `SelectableNodeViewModel` would then have some sort of `getSearchableString` method, which gets overridden appropriately by `ExerciseGroupViewModel`, `ExerciseViewModel`, and `SubmissionResultViewModel`.

Reply:
Hmm, apparently `TreeView` has a static `getViewModel` method (probably specifically because of this reason), could that be used here?

Reply:
It throws IllegalArgumentException

Reply:
`TreeView.getViewModel(x.getLastPathComponent())` shouldn't throw

Reply:
Indeed you're right. I will switch to this because it makes the code look better, however it does not really increase the type safety anyway because the implementation of this function is essentially this:

Reply:
![Screenshot_127](https://user-images.githubusercontent.com/73069541/106513477-b7d90280-64db-11eb-8f15-6064c6ce685e.png)

Reply:
You are correct that it doesn't improve type safety, but that's also not the concern here.

The concern is that with the current implementation, `ExercisesView` assumes that `TreeView` uses `DefaultMutableTreeNode`, but that is just an implementation detail in `TreeView` and not part of its public interface. In other words, it should be possible to change `DefaultMutableTreeNode` to something else without breaking any code using `TreeView`.

That is where `getViewModel` comes in. It encapsulates the implementation detail and now if the implementation of `TreeView` is modified, it won't break `ExercisesView`.